OpenRA.Game/Network/Connection.cs

```diff
@@ -30,12 +34,63 @@ public interface IConnection : IDisposable
 	{
 		int LocalClientId { get; }
 		ConnectionState ConnectionState { get; }
+		IPEndPoint EndPoint { get; }
+		string ErrorMessage { get; }
 		void Send(int frame, List<byte[]> orders);
 		void SendImmediate(IEnumerable<byte[]> orders);
 		void SendSync(int frame, byte[] syncData);
 		void Receive(Action<int, byte[]> packetFn);
 	}
 
+	public class ConnectionTarget
+	{
+		readonly DnsEndPoint[] endpoints;
+
+		public ConnectionTarget()
+		{
+			endpoints = new[] { new DnsEndPoint("invalid", 0) };
+		}
+
+		public ConnectionTarget(string host, int port)
+		{
+			endpoints = new[] { new DnsEndPoint(host, port) };
+		}
+
+		public ConnectionTarget(IEnumerable<DnsEndPoint> endpoints)
+		{
+			this.endpoints = endpoints.ToArray();
+			if (this.endpoints.Length == 0)
+			{
+				throw new ArgumentException("ConnectionTarget must have at least one address.");
+			}
+		}
+
+		public IEnumerable<IPEndPoint> GetConnectEndPoints()
+		{
+			return endpoints
+				.SelectMany(e =>
+				{
+					try
+					{
+						return Dns.GetHostAddresses(e.Host)
+							.Select(a => new IPEndPoint(a, e.Port));
+					}
+					catch (Exception)
+					{
+						return Array.Empty<IPEndPoint>();
```

Comment:
Style nit: Please use `Enumerable.Empty<IPEndPoint>()` for consistency with the rest of our code.